### PR TITLE
 Simplify IpRoyaltyVault to Support Claiming Revenue Without Snapshots

### PR DESCRIPTION
**Pull request by kingster-will**
## Description

This PR simplifies the `IpRoyaltyVault` contract by removing the requirement to use snapshots for claiming revenue. The new approach keeps track of the latest pending claimable revenue, allowing users to check the current pending revenue or claim all pending revenue with a single call.

## Key Changes

- **Remove Snapshot Functionality**: Eliminated the need for creating and managing snapshots.
- **Simplified Claim Process**: Users can now claim all pending revenue without specifying snapshot IDs.

### CODE REVIEW

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 1]
```diff
@@ -394,4 +399,78 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
...
+        if (balanceOf(from) < amount) revert Errors.IpRoyaltyVault__InsufficientBalance(address(this), from, amount);
+
+        IpRoyaltyVaultStorage storage $ = _getIpRoyaltyVaultStorage();
+        address[] memory tokenList = $.tokens.values();
+        for (uint256 i = 0; i < tokenList.length; i++) {
+            address token = tokenList[i];
+            _clearPendingRewards(from, token);
+            _clearPendingRewards(to, token);
+            $.userRewardInfo[token][to] = ($.poolInfo[token] * (balanceOf(to) + amount)) / totalSupply();
+            $.userRewardInfo[token][from] = ($.poolInfo[token] * (balanceOf(from) - amount)) / totalSupply();
```
Comment (DonFungible) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:450]: save gas by caching totalSupply in memory to reduce storage calls

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 2]
```diff
@@ -394,4 +399,78 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
...
+        if (ROYALTY_MODULE.isIpRoyaltyVault(claimer) && msg.sender != claimer)
+            revert Errors.IpRoyaltyVault__VaultsMustClaimAsSelf();
+
+        if (IP_ASSET_REGISTRY.isWhitelistedGroupRewardPool(claimer) && msg.sender != GROUPING_MODULE)
+            revert Errors.IpRoyaltyVault__GroupPoolMustClaimViaGroupingModule();
+
+        uint256[] memory claimableAmounts = new uint256[](tokenList.length);
+        for (uint256 i = 0; i < tokenList.length; i++) {
+            claimableAmounts[i] = _clearPendingRewards(claimer, tokenList[i]);
+            if (claimableAmounts[i] == 0) revert Errors.IpRoyaltyVault__NoClaimableTokens();
```
Comment (sebsadface) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:418]: Maybe better to just skip to the next token in the token list instead of reverting?

Comment (sebsadface) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:418]: Is it better to just skip to the next token in the tokenList instead of reverting?

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 1 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: Yes, good catch. totalSupply() should be cached in memory.
The PR was a draft for quick demonstrating the new simplified logic, not yet polished.
The totalSupply has been cached in the latest commit.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 3]
```diff
@@ -394,4 +399,78 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
...
+            $.userRewardInfo[tokenList[i]][claimer] += claimableAmounts[i];
+
+            $.claimVaultAmount[tokenList[i]] -= claimableAmounts[i];
+
```
[the remark this answers is not in the document]
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: the balance in the poolInfo is increasing only.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 2 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: It just following the existing logic of vault, the PR would not change this logic.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 4]
```diff
@@ -21,21 +17,23 @@ import { Errors } from "../../../lib/Errors.sol";
...
+    using SafeERC20 for IERC20;
 
     /// @dev Storage structure for the IpRoyaltyVault
     /// @param ipId The ip id to whom this royalty vault belongs to
-    /// @param lastSnapshotTimestamp The last snapshotted timestamp
-    /// @param pendingVaultAmount Amount of revenue token pending to be snapshotted
-    /// @param claimVaultAmount Amount of revenue token in the claim vault
-    /// @param claimableAtSnapshot Amount of revenue token claimable at a given snapshot
-    /// @param isClaimedAtSnapshot Indicates whether the claimer has claimed the revenue tokens at a given snapshot
+    /// @param lastSnapshotTimestamp [DEPRECATED] The last snapshotted timestamp
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:29]: Do we need [DEPRECATED] vs eliminate? Ideally we deploy cleanly at the start

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 5]
```diff
@@ -114,7 +114,6 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
         _mint(rtReceiver, supply);
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:114]: we may not need this line as variable was deprecated:

 `$.lastSnapshotTimestamp = uint40(block.timestamp);`

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 6]
```diff
@@ -280,112 +180,107 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
...
-    /// @return The amount of revenue token claimable
-    function _claimableRevenue(address account, uint256 snapshotId, address token) internal view returns (uint256) {
         IpRoyaltyVaultStorage storage $ = _getIpRoyaltyVaultStorage();
+        address[] memory tokenList = $.tokens.values();
+        uint256 totalSupply = totalSupply();
+        for (uint256 i = 0; i < tokenList.length; i++) {
+            address token = tokenList[i];
+            _clearPendingRewards(from, token);
+            _clearPendingRewards(to, token);
+            $.claimerInfo[token][to] = ($.poolInfo[token] * (balanceOf(to) + amount)) / totalSupply;
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:238]: It seems that updating `claimerInfo` always has to follow as-is the `_clearPendingRewards` call.

Therefore, would suggest moving `claimerInfo` updating as part of the function `_clearPendingRewards`. 

The function _clearPendingRewards could then be renamed to `claimPendingRewards`.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 7]
```diff
@@ -3,15 +3,11 @@ pragma solidity 0.8.26;
 
 import { EnumerableSet } from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
 import { PausableUpgradeable } from "@openzeppelin/contracts-upgradeable/utils/PausableUpgradeable.sol";
-// solhint-disable-next-line max-line-length
-import { ReentrancyGuardUpgradeable } from "@openzeppelin/contracts-upgradeable-v4/security/ReentrancyGuardUpgradeable.sol";
-// solhint-disable-next-line max-line-length
-import { ERC20SnapshotUpgradeable } from "@openzeppelin/contracts-upgradeable-v4/token/ERC20/extensions/ERC20SnapshotUpgradeable.sol";
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:9]: Think we no longer use @openzeppelin/contracts-upgradeable-v4/ - would double check but if so - it could be removed from dependencies and yarn.lock

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 8]
```diff
@@ -280,112 +180,107 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
...
+            return;
+        }
+        if (from == to) revert Errors.IpRoyaltyVault__SameFromToAddress(address(this), from);
+        // when transferring RoyaltyTokens (Vault) from a user to another user:
+        // 1. clear pending rewards of the (from) user
+        // 2. clear pending rewards of the (to) another user, if pending rewards of another user is not zero
+        // 3. update the rewardDebt of the (to) another user to accBalancePerShare * userAmount
+        // 4. update the rewardDebt of the (from) user to accBalancePerShare * userAmount
+        // 5. transfer the RoyaltyTokens (Vault) from the (from) user to the (to) another user
+        if (balanceOf(from) == 0) revert Errors.IpRoyaltyVault__ZeroBalance(address(this), from);
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:228]: `balanceOf(from)` call reads state and is being used multiple times in this function. We can store it in memory variable to save gas

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 9]
```diff
@@ -280,112 +180,107 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
...
-        // if the ip is tagged, then the unclaimed royalties are unavailable until the dispute is resolved
-        if (DISPUTE_MODULE.isIpTagged($.ipId)) return 0;
+        super._update(from, to, amount);
+    }
 
-        uint256 balance = balanceOfAt(account, snapshotId);
-        uint256 totalSupply = totalSupplyAt(snapshotId);
-        uint256 claimableToken = $.claimableAtSnapshot[snapshotId][token];
-        return $.isClaimedAtSnapshot[snapshotId][account][token] ? 0 : (balance * claimableToken) / totalSupply;
+    function _clearPendingRewards(address claimer, address token) internal returns (uint256 pending) {
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:245]: Can add natspec on all functions that don't have it yet for consistency with the rest of the contract

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 10]
```diff
@@ -280,112 +180,107 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
         }
     }
 
-    /// @notice Allows to claim revenue tokens on behalf of the ip royalty vault by snapshot batch
-    /// @param snapshotIds The list of snapshot ids
+    /// @notice Get total amount of revenue token claimable by a token holder
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:183]: Get total amount of revenue token claimable by a token holder

could be for clarity

Get total amount of revenue token claimable by a royalty token holder

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 11]
```diff
@@ -280,112 +180,107 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
...
+    /// @param claimer The address of the token holder
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:184]: The address of the token holder

could be

The address of the royalty token holder

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 12]
```diff
@@ -280,112 +180,107 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
...
+        for (uint256 i = 0; i < tokenList.length; i++) {
+            address token = tokenList[i];
+            _clearPendingRewards(from, token);
+            _clearPendingRewards(to, token);
+            $.claimerInfo[token][to] = ($.poolInfo[token] * (balanceOf(to) + amount)) / totalSupply;
+            $.claimerInfo[token][from] = ($.poolInfo[token] * (balanceOf(from) - amount)) / totalSupply;
+        }
 
-        // if the ip is tagged, then the unclaimed royalties are unavailable until the dispute is resolved
-        if (DISPUTE_MODULE.isIpTagged($.ipId)) return 0;
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:368]: One of the requirements as penalty of a dispute is the inability to claim revenue tokens from the vault when an IP is tagged.

Before we had this line:
`if (DISPUTE_MODULE.isIpTagged($.ipId)) return 0;`

In the new version we want to achieve the same requirement. Now for where to place this - if on transfer - would say that if we may want to measure the gas cost to confirm if it's ok to add it there or if another solution is preferrable.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 13]
```diff
@@ -280,112 +180,107 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
...
 
-    /// @notice A function to calculate the amount of revenue token claimable by a token holder at certain snapshot
-    /// @param account The address of the token holder
-    /// @param snapshotId The snapshot id
-    /// @param token The revenue token to claim
-    /// @return The amount of revenue token claimable
-    function _claimableRevenue(address account, uint256 snapshotId, address token) internal view returns (uint256) {
         IpRoyaltyVaultStorage storage $ = _getIpRoyaltyVaultStorage();
+        address[] memory tokenList = $.tokens.values();
+        uint256 totalSupply = totalSupply();
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:233]: `uint256 totalSupply = totalSupply();` reads state and is a step where we could save on gas by replacing with an immutable variable of the vault

Something along the lines of:

function totalSupply() public view override returns (uint256) {
        return TOTAL_SUPPLY;
   }

And then can eliminate `uint256 totalSupply = totalSupply();` and use `TOTAL_SUPPLY` directly

Given all vaults have the same total supply and it stays constant over the lifetime of the vault (no minting and no burning) - this is possible to do to save gas on every royalty token transfer.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 14]
```diff
@@ -280,112 +180,107 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
...
+            claimedAmounts[i] = _clearPendingRewards(claimer, tokenList[i]);
+            if (claimedAmounts[i] == 0) revert Errors.IpRoyaltyVault__NoClaimableTokens();
+            $.claimerInfo[tokenList[i]][claimer] += claimedAmounts[i];
+        }
+
+        return claimedAmounts;
+    }
+
+    function _claimableRevenue(address claimer, address token) internal view returns (uint256) {
+        // accBalance // accumulate revenue tokens in the vault
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:273]: // accBalance - accumulated revenue tokens in the vault

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 15]
```diff
@@ -280,112 +180,107 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
...
+        // totalSupply = totalSupply() // totalSupply of RoyaltyTokens of the vault (IpRoyaltyVault)
+        // accBalancePerShare = accBalance / totalSupply
+        // userAmount = balanceOf(user) // user amount of RoyaltyTokens (IpRoyaltyVault),
+        // means how many share the user has
+        // pending = (accBalancePerShare * userAmount) - userRewardInfo[token][user].rewardDebt
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:278]: claimerInfo vs userRewardInfo

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 16]
```diff
@@ -280,112 +180,107 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
...
+
+        return claimedAmounts;
+    }
+
+    function _claimableRevenue(address claimer, address token) internal view returns (uint256) {
+        // accBalance // accumulate revenue tokens in the vault
+        // totalSupply = totalSupply() // totalSupply of RoyaltyTokens of the vault (IpRoyaltyVault)
+        // accBalancePerShare = accBalance / totalSupply
+        // userAmount = balanceOf(user) // user amount of RoyaltyTokens (IpRoyaltyVault),
+        // means how many share the user has
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:277]: means how many share the user has

could be

means how much share the user has

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 17]
```diff
@@ -280,112 +180,107 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
...
+        }
+
+        return claimedAmounts;
+    }
+
+    function _claimableRevenue(address claimer, address token) internal view returns (uint256) {
+        // accBalance // accumulate revenue tokens in the vault
+        // totalSupply = totalSupply() // totalSupply of RoyaltyTokens of the vault (IpRoyaltyVault)
+        // accBalancePerShare = accBalance / totalSupply
+        // userAmount = balanceOf(user) // user amount of RoyaltyTokens (IpRoyaltyVault),
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:276]: remove ","

File: contracts/lib/Errors.sol [hunk 18]
```diff
@@ -699,6 +699,14 @@ library Errors {
     /// @notice Group reward pool must claim via GroupingModule.
     error IpRoyaltyVault__GroupPoolMustClaimViaGroupingModule();
```
Comment (Spablob) [inline contracts/lib/Errors.sol:701]: On the Errors.sol file - there are likely errors that are no longer being used and can be deleted

File: contracts/modules/royalty/policies/VaultController.sol [hunk 19]
```diff
@@ -23,14 +23,6 @@ abstract contract VaultController is IVaultController, ProtocolPausableUpgradeab
     bytes32 private constant VaultControllerStorageLocation =
         0x88cf5a7bd03e240c4fc740fb2d1a8664ec6fa4816f867d60f968080755fb1700;
 
-    /// @dev Set the snapshot interval
```
Comment (Spablob) [inline contracts/modules/royalty/policies/VaultController.sol:26]: Can remove (or DEPRECATE) `uint256 snapshotInterval;` from the variables on VaultController.sol

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 5 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol:113]: removed the line.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 6 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: For 1), as we need to update `claimerInfo` when transfer RTs. even if `_claimableRevenue` is zero. and with latest update, it update `claimerInfo` without always follow `_clearPendingRewards` call. it is no need to move it into the _clearPendingRewards.
For 2), good point, rename it to `_claimPendingRevenue()` for consistency.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 7 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol:9]: Yes, removed from yarn.lock. 
and the PR has already removed it from dependencies in package.json.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 20]
```diff
@@ -280,112 +180,107 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
...
-    /// @param account The address of the token holder
-    /// @param snapshotId The snapshot id
-    /// @param token The revenue token to claim
-    /// @return The amount of revenue token claimable
-    function _claimableRevenue(address account, uint256 snapshotId, address token) internal view returns (uint256) {
         IpRoyaltyVaultStorage storage $ = _getIpRoyaltyVaultStorage();
+        address[] memory tokenList = $.tokens.values();
+        uint256 totalSupply = totalSupply();
+        for (uint256 i = 0; i < tokenList.length; i++) {
+            address token = tokenList[i];
```
[the remark this answers is not in the document]
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: using tokenList[i] directly.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 8 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: yes, good point. store balance into local var.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 9 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: added comments to all internal functions.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 10 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: changed

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 11 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: changed.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 13 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol:234]: Good point!
created an issue to track this enhancement: #294. The change will not be trivial, as it requires modifications to the RoyaltyModule to ensure they share the same configuration. Additionally, updates to the DeployHelper may also be necessary. Therefore, it would be better to handle this in a separate PR.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 14 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: updated.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 15 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: updated.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 16 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: updated.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 17 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: updated.

File: contracts/modules/royalty/policies/VaultController.sol (hunk 19 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/VaultController.sol:26]: yes, mark it as DEPRECATED.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 21]
```diff
@@ -44,7 +44,7 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20Upgradeable, ReentrancyGuardUpg
         mapping(uint256 snapshotId => mapping(address claimer => mapping(address token => bool))) isClaimedAtSnapshot;
         EnumerableSet.AddressSet tokens;
         mapping(address token => uint256 accBalance) poolInfo;
-        mapping(address token => mapping(address claimer => uint256 revenueDebt)) claimerInfo;
+        mapping(address token => mapping(address claimer => int256 revenueDebt)) claimerInfo;
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:47]: why the change from uint256 to int256?

File: contracts/lib/Errors.sol (hunk 18 above)
Comment (kingster-will) [reply contracts/lib/Errors.sol:695]: I see. removed unused errors.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 22]
```diff
@@ -233,10 +233,14 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20Upgradeable, ReentrancyGuardUpg
...
-            _clearPendingRewards(to, token);
-            $.claimerInfo[token][to] = ($.poolInfo[token] * (balanceOf(to) + amount)) / totalSupply;
-            $.claimerInfo[token][from] = ($.poolInfo[token] * (balanceOf(from) - amount)) / totalSupply;
+            uint256 pendingFrom = _claimableRevenue(from, token);
+            uint256 pendingTo = _claimableRevenue(to, token);
+            $.claimerInfo[token][to] =
+                int256(($.poolInfo[token] * (balanceOf(to) + amount)) / totalSupply) -
+                int256(pendingTo);
+            $.claimerInfo[token][from] =
+                int256(($.poolInfo[token] * (balanceOf(from) - amount)) / totalSupply) -
```
[the remark this answers is not in the document]
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: good point! 
stored in a local variable.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 21 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol:47]: it because the `revenueDebt` could be negative value which indicate there is pending revenue of claimer.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 23]
```diff
@@ -233,10 +233,14 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20Upgradeable, ReentrancyGuardUpg
...
-            _clearPendingRewards(from, token);
-            _clearPendingRewards(to, token);
-            $.claimerInfo[token][to] = ($.poolInfo[token] * (balanceOf(to) + amount)) / totalSupply;
-            $.claimerInfo[token][from] = ($.poolInfo[token] * (balanceOf(from) - amount)) / totalSupply;
+            uint256 pendingFrom = _claimableRevenue(from, token);
+            uint256 pendingTo = _claimableRevenue(to, token);
+            $.claimerInfo[token][to] =
+                int256(($.poolInfo[token] * (balanceOf(to) + amount)) / totalSupply) -
+                int256(pendingTo);
+            $.claimerInfo[token][from] =
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol]: $.claimerInfo[tokenList[i]][from] =
                int256(($.poolInfo[tokenList[i]] * (balanceOfFrom - amount)) / totalSupply) -
                int256(pendingFrom);

Will it revert on royalty token transfer if the user has not claimed yet?

Because,
($.poolInfo[tokenList[i]] * (balanceOfFrom - amount)) / totalSupply) < pendingFrom

Thinking of the DEX contract that holds the royalty tokens for swapping - would it grind to a halt? Or am I missing something?

Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: It would out revert, since the value type of `claimerInfo` is `int256` which hold negative value.
Such case that transfer RTs while still have pending revenue, has also been covered in test case `test_IpRoyaltyVault_payRev_then_transferRTs`:
https://github.com/storyprotocol/protocol-core-v1/pull/291/files#diff-7b1113b9f56f60a900b64a18ab4e5962438cba8524418eda505d359e8c83c0cfR533

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 4 above)
Comment (Spablob) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol:29]: Noticed that the variables were eliminated but the natspec still seems to have the [DEPRECATED] variables?

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 24]
```diff
@@ -133,132 +132,34 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
...
-            revert Errors.IpRoyaltyVault__InsufficientTimeElapsedSinceLastSnapshot();
-
-        uint256 snapshotId = _snapshot();
-        $.lastSnapshotTimestamp = uint40(block.timestamp);
-
-        uint256 noRevenueCounter;
-        address[] memory tokenList = $.tokens.values();
-        for (uint256 i = 0; i < tokenList.length; i++) {
-            if (IERC20Upgradeable(tokenList[i]).balanceOf(address(this)) == 0) {
-                $.tokens.remove(tokenList[i]);
```
[the remark this answers is not in the document]
Comment (Spablob) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol:151]: Why do we need to track revenue tokens with balance zero? What breaks if we remove the token from the token list?

Can't we have something along the lines of

```
if (IERC20Upgradeable(tokenList[i]).balanceOf(address(this)) == 0) {
                $.tokens.remove(tokenList[i]);
```

inside the function `_claimRevenueOnBehalf` ?

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 25]
```diff
@@ -45,6 +43,8 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
         mapping(uint256 snapshotId => mapping(address token => uint256 amount)) claimableAtSnapshot;
         mapping(uint256 snapshotId => mapping(address claimer => mapping(address token => bool))) isClaimedAtSnapshot;
         EnumerableSet.AddressSet tokens;
+        mapping(address token => uint256 accBalance) poolInfo;
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:46]: can expose `poolInfo` and `claimerInfo` variables in external view function like ipId() and tokens()

Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:46]: would rename `poolInfo` to `vaultInfo` given it's not a pool

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 26]
```diff
@@ -21,21 +17,23 @@ import { Errors } from "../../../lib/Errors.sol";
...
-    /// @param claimVaultAmount Amount of revenue token in the claim vault
-    /// @param claimableAtSnapshot Amount of revenue token claimable at a given snapshot
-    /// @param isClaimedAtSnapshot Indicates whether the claimer has claimed the revenue tokens at a given snapshot
+    /// @param lastSnapshotTimestamp [DEPRECATED] The last snapshotted timestamp
+    /// @param pendingVaultAmount [DEPRECATED] Amount of revenue token pending to be snapshotted
+    /// @param claimVaultAmount [DEPRECATED] Amount of revenue token in the claim vault
+    /// @param claimableAtSnapshot [DEPRECATED] Amount of revenue token claimable at a given snapshot
+    /// @param isClaimedAtSnapshot [DEPRECATED] Indicates whether the claimer has claimed the token at a given snapshot
     /// @param tokens The list of revenue tokens in the vault
+    /// @param poolInfo The accumulated balance of revenue tokens in the vault
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:35]: Should we rename the two new variables away from "Info"?

For example
`vaultInfo` rename to `vaultAccBalance`
`claimerInfo` rename to `revenueTokenDebt`

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 27]
```diff
@@ -21,21 +17,23 @@ import { Errors } from "../../../lib/Errors.sol";
...
+    /// @param claimerInfo The revenue debt of the claimer
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:36]: Would add to this comment/natspec what is the meaning of a positive and a negative debt

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 4 above)
Comment (Spablob) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol:29]: Got it, we can add an issue to remove all [DEPRECATED] variables for fresh deployment

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 25 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: Added view functions `revenueDebt` and `vaultAccBalances`.

Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: renamed to `vaultAccBalances`

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 26 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: renamed.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 27 above)
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol]: added more comments.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol (hunk 24 above)
[the remark this answers is not in the document]
Comment (kingster-will) [reply contracts/modules/royalty/policies/IpRoyaltyVault.sol:151]: We need to keep the token in the token list even when its balance is zero because removing it could prevent us from properly adjusting a user’s `revenueDebt` for that token when they transfer RTs to another user, as the token would no longer be listed. Since the token’s balance can be increased later, the user’s claimable revenue for that token might become inaccurate if it was previously removed from the list.

File: contracts/modules/royalty/policies/IpRoyaltyVault.sol [hunk 28]
```diff
@@ -207,7 +207,7 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20Upgradeable, ReentrancyGuardUpg
     /// @param claimer The address of the claimer
     /// @param token The revenue token to check
     /// @return The revenue debt of the claimer
-    function claimerRevenueDebt(address claimer, address token) external view returns (int256) {
+    function revenueDebt(address claimer, address token) external view returns (int256) {
```
Comment (Spablob) [inline contracts/modules/royalty/policies/IpRoyaltyVault.sol:210]: can we have the same name for variable and for view function name for consistency?

right now `revenueDebt` in view function and `claimerRevenueDebt` in variable. 

Could be either one as long as they are the same for consistency

**Review by Spablob — APPROVED** [review]
